packages/gatsby/src/redux/actions/public.js

```diff
@@ -1182,6 +1192,65 @@ actions.createJob = (job: Job, plugin?: ?Plugin = null) => {
   }
 }
 
+/**
+ * Create a "job". This is a long-running process that are generally
+ * started as side-effects to GraphQL queries.
+ * [`gatsby-plugin-sharp`](/packages/gatsby-plugin-sharp/) uses this for
+ * example.
+ *
+ * Gatsby doesn't finish its bootstrap until all jobs are ended.
+ * @param {Object} job A job object with name, inputPaths, outputDir and args
+ * @param {string} job.name The name of the job you want to execute
+ * @param {string[]} job.inputPaths The inputPaths that are needed to run
+ * @param {string} job.outputDir The directory where all files are being saved to
+ * @param {Object} job.args The arguments the job needs to execute
+ * @returns {Promise<unknown>} Promise to see if the job is done executing
+ * @example
+ * createJobV2({ name: `IMAGE_PROCESSING`, inputPaths: [`something.jpeg`], outputDir: `public/static`, args: { width: 100, height: 100 } })
+ */
+actions.createJobV2 = (job: JobV2, plugin: Plugin) => (dispatch, getState) => {
+  const currentState = getState()
+  const internalJob = createInternalJob(
+    job,
+    plugin,
+    currentState.program.directory
+  )
+
+  // Check if we already ran this job before, if yes we return the result
+  if (
+    currentState.jobsV2 &&
+    currentState.jobsV2.done.has(internalJob.contentDigest)
+  ) {
+    return Promise.resolve(
+      currentState.jobsV2.done.get(internalJob.contentDigest).result
+    )
+  }
```

Comment:
not sure if common - but she we similarly check in progress jobs and return its promise if there is in same progress job?

Reply:
we do this in enqueuJob
https://github.com/gatsbyjs/gatsby/pull/19858/files/225212a533052e16d5fc6a925004fee39e5c608a#diff-af5cc3543290053211e5858a2171132bR171-R173

Reply:
I assume this is by usage of `jobsInProcess` lookup?

Reply:
I don't think I understand the question. The done queue are all jobs ever run so we can return a result if they have any, like sqip or any expensive operation.

Reply:
Just the fact the we do those lookups (for jobs that finished + jobs that are in progress) in different places (one in action creator and one in jobs-manager) seems unintuitive. Can we move both of those lookups to single module or is there a reason.

For reference this is how we do something similiar in loader.js and both done + in progress are near each other, so it's easier do understand what's happening: https://github.com/gatsbyjs/gatsby/blob/e854c1b078aff1f5ff40356580bd5046f58b1aa7/packages/gatsby/cache-dir/loader.js#L167-L173

Reply:
It depends on how you look at it. Jobs manager is in control of everything that's happening in the current run. Redux is in control of what has been running before. If we remove redux our job manager will still not run jobs twice in the current runtime.

That's the reason why it's scattered (I also hate redux because of this). The only reason why inProgress is exported is to be compatible with v1 so moving it here seems wrong to me.

---

After reading your other comment about Jobs in progress not being removed, it might make sense to move it here after all and let the jobs manager just run jobs. I first had written everything separately as I didn't have anything in redux. I don't like the architecture, but we use redux everywhere so I guess to keep things consistent.

Reply:
After the weekend I do want to keep it like it is. I don't want to expose the `jobsInProcess` map outside of the jobs manager. I only do this to be v1 compat and would like to keep it that way.

Reply:
That's fine not to swap code around - maybe just some comment that "in flight" jobs are checked by job manager?